src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -225,28 +240,39 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
                 }
             }
 
-            private Task GetTask(Project project, Func<Task> func, CancellationToken cancellationToken)
-            {
-                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
-                return isRemoteWorkspace
-                    ? Task.Run(func, cancellationToken)
-                    : func();
-            }
-
-            private Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/33172", AllowGenericEnumeration = false, AllowCaptures = false)]
+            private Task UpdateReferencesAsync(Project project, CancellationToken cancellationToken)
             {
                 // Process all metadata references. If it remote workspace, do this in parallel.
-                var tasks = new List<Task>();
+                var tasks = new List<Task>(project.MetadataReferences.Count);
+                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
 
-                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
+                foreach (var metadataReference in project.MetadataReferences.ToImmutableArrayOrEmpty())
                 {
-                    tasks.Add(
-                        GetTask(project, () => UpdateReferenceAsync(project, reference, cancellationToken), cancellationToken));
+                    if (metadataReference is PortableExecutableReference portableExecutableReference)
+                    {
+                        var task = isRemoteWorkspace ?
+                            UpdateReferenceInServiceProcessAsync(project, portableExecutableReference, cancellationToken) :
+                            UpdateReferenceInCurrentProcessAsync(project, portableExecutableReference, cancellationToken);
+
+                        tasks.Add(task);
+                    }
                 }
 
                 return Task.WhenAll(tasks);
             }
 
+            private Task UpdateReferenceInCurrentProcessAsync(Project project, PortableExecutableReference reference, CancellationToken cancellationToken)
+            {
+                return UpdateReferenceAsync(project, reference, cancellationToken);
+            }
+
+            private Task UpdateReferenceInServiceProcessAsync(Project project, PortableExecutableReference reference, CancellationToken cancellationToken)
+            {
+                return Task.Run(() => UpdateReferenceAsync(project, reference, cancellationToken), cancellationToken);
```

Comment:
won't this allocate as well?  i'm not sure why this is better.

Reply:
Because we don't create the closure for the in-process version

Reply:
I've retained perf characteristics for VSWindows+remote workspace and improved them for in-process

Reply:
But the in-process version is going away.  see https://github.com/dotnet/roslyn/pull/27983.  So will that really be super helpful?

Reply:
We don't have ServiceHub available, so how is that completely going away?

Reply:
I suppose opt-ing in is going away the ability to run without servicehub still exists.

Reply:
Reading #27983, it seems like that's the case only when OOP _is_ enabled. This came up when profiling VSMac, so I'm optimizing that scenario (in-process), because fixing that for the OOP scenario is impossible (Task.Run has icky API).